### ISSUE TITLE
Getting 403 error for STS: GetFederationToken when using credential_process

### ISSUE THREAD

**Issue by TryTryAgain** [common-fate/granted#151]
I see a few related issues, sorry if this issue is redundant... Although I'm not using SAML, I'm using `credential_process` to retrieve sts credentials from a remote machine/instance-profile role. 

`operation error STS: GetFederationToken, https response error StatusCode: 403, RequestID: 5d029b3b-9e3a-46dd-b043-1f4e50fb06ec, api error AccessDenied: Cannot call GetFederationToken with session credentials`

**Comment by JoshuaWilkes** [comment]
I believe that the IAM credentials you are using do not have an IAM policy which grants access to call `sts:GetFederationToken`

The credentials your are retrieving need to be able to call this API, an example IAM policy statement which allows this API to be called.

{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "sts:GetFederationToken",
      "Resource": "*"
    },
]
}

**Comment by TryTryAgain** [comment]
@JoshuaWilkes  I'm using AdministratorAccess granted from a cross-account admin role. STS credentials are sourced via a credential_process that uses a remote machine's Instance profile role. 

My ~/.aws/config and ~/.aws/credentials had to change slightly to work with granted but it works and worked before; as I mentioned https://github.com/common-fate/granted/issues/117#issuecomment-1105208389 too.

**Comment by TryTryAgain** [comment]
So... I found this https://github.com/aws/aws-sdk-php/issues/1876#issuecomment-537152907 ...and I guess:

"As the error message suggests, session credentials cannot be used to initialize a new STS session via a getSessionToken call. Credentials that include a Token value (e.g. STS temporary security credentials, assumed IAM roles, instance profile credentials) are considered session credentials and thus cannot be used to obtain a new session token via a getSessionToken call."

But it seems like maybe there is a way to use the "proper" credentials and not the session credentials? I'm really lost and not sure why this wouldn't work, all new to me. Would love to get it working though, any help is appreciated.

**Comment by holly-evans** [comment]
Hi @TryTryAgain, I was wrestling with something sort-of similar. One of my accounts uses long-lived IAM keys that I don't want to store in plain-text, so I use aws-vault and call them using credential_process. I couldn't pull from aws-vault directly or use session credentials so I needed to call `aws sts get-federation-token ...` myself in the credential process.

You're correct, you cannot use session credentials to generate more session credentials - a safety feature since otherwise you could perpetually make new sessions off one temporary session.

Your error is for `GetFederationToken`, so I'm assuming you were trying to open the console with `assume -c <profile>`? Could you share what your credential_process looks like? What is your goal with Granted in this scenario since you already have temporary session credentials?

**Comment by TryTryAgain** [comment]
Right, this has been a learning experience. The key is "Ultimately, if only credentials you have access to are temporary session credentials without a role, the console part will not work."... so we will need to issue a long-living key (I suppose we'll rotate it daily) and that will do. Thanks for the explanation and examples. I'm sure once I get that to be the source of the role switching credentials all will be working as expected as you said.
